Switch fall-through unsupported

**If eligible for the bounty, I wish this to be donated to a local animal charity of the maintainer's choice**


#### GCC version:
```bash
gcc (SUSE Linux) 10.2.1 20201202 [revision e563687cf9d3d1278f45aaebd03e0f66531076c9]
```

#### clang version:
```bash
clang version 11.0.1
```

#### Output of `gcc -Werror -Wall -Wextra -Wpedantic -pedantic -std=c11 -c $FNAME`
```bash
(return code 0)
```

#### Output of `clang -Werror -Wall -Wextra -Wpedantic -pedantic -std=c11 -c $FNAME`
```bash
(return code 0)
```

#### Output of `cnip $FNAME`
```bash
uninit-exhaustive-switch-bug.c:5:2 error: `case' label not within a switch statement
  case 1:
  ^

(return code 0)
```

#### Content of file
```c
void test(int i) {
  switch (i) {
  case 0:
  case 1:
    break;
  }
}
```

#### `psychec` version
```bash
commit 6d7b8249e0f28af105112b217cb33a24f666e148
Author: Leandro T. C. Melo <LTCMELO@GMAIL.COM>
Date:   Tue Feb 23 07:47:25 2021 -0300

    link expression-parsing table in Doxygen documentation
```

--- a/C/SyntaxTree.cpp
+++ b/C/SyntaxTree.cpp
@@ -204,7 +204,7 @@ void SyntaxTree::buildTree(SyntaxCategory syntaxCat)
 
         case SyntaxCategory::Statements: {
             StatementSyntax* stmt = nullptr;
-            parser.parseStatement(stmt);
+            parser.parseStatement(stmt, Parser::StatementContext::None);
             P->rootNode_ = stmt;
             break;
          }

--- a/C/parser/Parser.h
+++ b/C/parser/Parser.h
@@ -196,7 +196,8 @@ private:
     enum class StatementContext : uint8_t
     {
         None,
-        WithinSwitch
+        WithinSwitch,
+        WithinLoop,
     };
 
     template <class NodeT, class... Args> NodeT* makeNode(Args&&... args) const;
@@ -376,24 +377,23 @@ private:
     //------------//
     // Statements //
     //------------//
-    bool parseStatement(StatementSyntax*& stmt,
-                        StatementContext stmtCtx = StatementContext::None);
+    bool parseStatement(StatementSyntax*& stmt, StatementContext stmtCtx);
     bool parseCompoundStatement_AtFirst(
             StatementSyntax*& stmt,
-            StatementContext stmtCtx = StatementContext::None);
+            StatementContext stmtCtx);
     bool parseDeclarationStatement(
             StatementSyntax*& stmt,
             bool (Parser::*parseDecl)(DeclarationSyntax*&));
     bool parseExpressionStatement(StatementSyntax*& stmt);
-    bool parseLabeledStatement_AtFirst(StatementSyntax*& stmt);
-    bool parseIfStatement_AtFirst(StatementSyntax*& stmt);
+    bool parseLabeledStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
+    bool parseIfStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
     bool parseSwitchStatement_AtFirst(StatementSyntax*& stmt);
-    bool parseWhileStatement_AtFirst(StatementSyntax*& stmt);
-    bool parseDoStatement_AtFirst(StatementSyntax*& stmt);
-    bool parseForStatement_AtFirst(StatementSyntax*& stmt);
+    bool parseWhileStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
+    bool parseDoStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
+    bool parseForStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
     bool parseGotoStatement_AtFirst(StatementSyntax*& stmt);
-    bool parseContinueStatement_AtFirst(StatementSyntax*& stmt);
-    bool parseBreakStatement_AtFirst(StatementSyntax*& stmt);
+    bool parseContinueStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
+    bool parseBreakStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx);
     bool parseReturnStatement_AtFirst(StatementSyntax*& stmt);
     bool parseExtGNU_AsmStatement(StatementSyntax*& stmt);
     void parseExtGNU_AsmQualifiers(SpecifierListSyntax*& specList);

--- a/C/parser/Parser_Declarations.cpp
+++ b/C/parser/Parser_Declarations.cpp
@@ -333,7 +333,7 @@ bool Parser::parseDeclarationOrFunctionDefinition_AtFollowOfSpecifiers(
                         decl = funcDef;
                         funcDef->specs_ = const_cast<SpecifierListSyntax*>(specList);
                         funcDef->decltor_ = decltor;
-                        parseCompoundStatement_AtFirst(funcDef->body_);
+                        parseCompoundStatement_AtFirst(funcDef->body_, StatementContext::None);
                         return true;
                     }
                 }

--- a/C/parser/Parser_Expressions.cpp
+++ b/C/parser/Parser_Expressions.cpp
@@ -223,7 +223,7 @@ bool Parser::parseExtGNU_StatementExpression_AtFirst(ExpressionSyntax *&expr)
     gnuExpr->openParenTkIdx_ = consume();
 
     StatementSyntax* statement = nullptr;
-    parseCompoundStatement_AtFirst(statement);
+    parseCompoundStatement_AtFirst(statement, StatementContext::None);
     if (statement->asCompoundStatement())
         gnuExpr->stmt_ = statement->asCompoundStatement();
     return matchOrSkipTo(CloseParenToken, &gnuExpr->closeParenTkIdx_);
@@ -499,7 +499,7 @@ bool Parser::parseExpressionWithPrecedencePostfix(ExpressionSyntax*& expr)
         return false;
 
     while (true) {
-        SyntaxKind exprK = Unknown;
+        SyntaxKind exprK = UnknownSyntax;
         switch (peek().kind()) {
             /* 6.5.2.1 */
             case OpenBracketToken: {
@@ -541,7 +541,7 @@ bool Parser::parseExpressionWithPrecedencePostfix(ExpressionSyntax*& expr)
                 [[fallthrough]];
 
             case ArrowToken: {
-                if (exprK == Unknown)
+                if (exprK == UnknownSyntax)
                     exprK = IndirectMemberAccessExpression;
                 if (!parsePostfixExpression_AtPostfix<MemberAccessExpressionSyntax>(
                             expr,
@@ -569,7 +569,7 @@ bool Parser::parseExpressionWithPrecedencePostfix(ExpressionSyntax*& expr)
                 [[fallthrough]];
 
             case MinusMinusToken: {
-                if (exprK == Unknown)
+                if (exprK == UnknownSyntax)
                     exprK = PostDecrementExpression;
                 if (!parsePostfixExpression_AtPostfix<PostfixUnaryExpressionSyntax>(
                             expr,

--- a/C/parser/Parser_Statements.cpp
+++ b/C/parser/Parser_Statements.cpp
@@ -84,7 +84,7 @@ bool Parser::parseStatement(StatementSyntax*& stmt, StatementContext stmtCtx)
 
         case IdentifierToken: {
             if (peek(2).kind() == ColonToken)
-                return parseLabeledStatement_AtFirst(stmt);
+                return parseLabeledStatement_AtFirst(stmt, stmtCtx);
 
             Backtracker BT(this);
             if (!parseExpressionStatement(stmt)) {
@@ -101,7 +101,7 @@ bool Parser::parseStatement(StatementSyntax*& stmt, StatementContext stmtCtx)
             return parseCompoundStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_if:
-            return parseIfStatement_AtFirst(stmt);
+            return parseIfStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_switch:
             return parseSwitchStatement_AtFirst(stmt);
@@ -109,30 +109,30 @@ bool Parser::parseStatement(StatementSyntax*& stmt, StatementContext stmtCtx)
         case Keyword_case:
             if (stmtCtx != StatementContext::WithinSwitch)
                 diagnosticsReporter_.UnexpectedCaseLabelOutsideSwitch();
-            return parseLabeledStatement_AtFirst(stmt);
+            return parseLabeledStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_default:
             if (stmtCtx != StatementContext::WithinSwitch)
                 diagnosticsReporter_.UnexpectedDefaultLabelOutsideSwitch();
-            return parseLabeledStatement_AtFirst(stmt);
+            return parseLabeledStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_while:
-            return parseWhileStatement_AtFirst(stmt);
+            return parseWhileStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_do:
-            return parseDoStatement_AtFirst(stmt);
+            return parseDoStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_for:
-            return parseForStatement_AtFirst(stmt);
+            return parseForStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_goto:
             return parseGotoStatement_AtFirst(stmt);
 
         case Keyword_continue:
-            return parseContinueStatement_AtFirst(stmt);
+            return parseContinueStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_break:
-            return parseBreakStatement_AtFirst(stmt);
+            return parseBreakStatement_AtFirst(stmt, stmtCtx);
 
         case Keyword_return:
             return parseReturnStatement_AtFirst(stmt);
@@ -227,7 +227,7 @@ void Parser::maybeAmbiguateStatement(StatementSyntax*& stmt)
 
     VariableAndOrFunctionDeclarationSyntax* varDecl = nullptr;
     varDecl = makeNode<VariableAndOrFunctionDeclarationSyntax>();
-    varDecl->specs_ =makeNode<SpecifierListSyntax>(typedefName);
+    varDecl->specs_ = makeNode<SpecifierListSyntax>(typedefName);
     varDecl->semicolonTkIdx_ = exprStmt->semicolonTkIdx_;
     varDecl->decltors_ = makeNode<DeclaratorListSyntax>(decltor);
 
@@ -261,7 +261,7 @@ bool Parser::parseCompoundStatement_AtFirst(StatementSyntax*& stmt,
     StatementListSyntax** stmtList_cur = &block->stmts_;
 
     while (true) {
-        StatementSyntax* insideStmt = nullptr;
+        StatementSyntax* innerStmt = nullptr;
         switch (peek().kind()) {
             case EndOfFile:
                 diagnosticsReporter_.ExpectedToken(CloseBraceToken);
@@ -272,7 +272,7 @@ bool Parser::parseCompoundStatement_AtFirst(StatementSyntax*& stmt,
                 return true;
 
             default: {
-                if (!parseStatement(insideStmt, stmtCtx)) {
+                if (!parseStatement(innerStmt, stmtCtx)) {
                     skipTo(CloseBraceToken);
                     continue;
                 }
@@ -280,7 +280,7 @@ bool Parser::parseCompoundStatement_AtFirst(StatementSyntax*& stmt,
             }
         }
 
-        *stmtList_cur = makeNode<StatementListSyntax>(insideStmt);
+        *stmtList_cur = makeNode<StatementListSyntax>(innerStmt);
         stmtList_cur = &(*stmtList_cur)->next;
     }
 
@@ -320,7 +320,7 @@ bool Parser::parseExpressionStatement(StatementSyntax*& stmt)
  *
  * \remark 6.8.4.1
  */
-bool Parser::parseIfStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseIfStatement_AtFirst(StatementSyntax*& stmt, StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == Keyword_if,
@@ -334,14 +334,14 @@ bool Parser::parseIfStatement_AtFirst(StatementSyntax*& stmt)
     if (!(match(OpenParenToken, &ifStmt->openParenTkIdx_)
             && parseExpression(ifStmt->cond_)
             && match(CloseParenToken, &ifStmt->closeParenTkIdx_)
-            && parseStatement(ifStmt->stmt_))) {
+            && parseStatement(ifStmt->stmt_, stmtCtx))) {
         return ignoreStatement();
     }
 
     if (peek().kind() == Keyword_else) {
         ifStmt->elseKwTkIdx_ = consume();
         return checkStatementParse(
-                    parseStatement(ifStmt->elseStmt_));
+                    parseStatement(ifStmt->elseStmt_, stmtCtx));
     }
     return true;
 }
@@ -375,7 +375,8 @@ bool Parser::parseSwitchStatement_AtFirst(StatementSyntax*& stmt)
  *
  * \remark 6.8.1
  */
-bool Parser::parseLabeledStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseLabeledStatement_AtFirst(StatementSyntax*& stmt,
+                                           StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == IdentifierToken
@@ -420,7 +421,7 @@ bool Parser::parseLabeledStatement_AtFirst(StatementSyntax*& stmt)
     }
 
     return checkStatementParse(
-                parseStatement(labelStmt->stmt_));
+                parseStatement(labelStmt->stmt_, stmtCtx));
 }
 
 /**
@@ -429,7 +430,8 @@ bool Parser::parseLabeledStatement_AtFirst(StatementSyntax*& stmt)
  *
  * \remark 6.8.5.1
  */
-bool Parser::parseWhileStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseWhileStatement_AtFirst(StatementSyntax*& stmt,
+                                         StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == Keyword_while,
@@ -444,7 +446,7 @@ bool Parser::parseWhileStatement_AtFirst(StatementSyntax*& stmt)
                 match(OpenParenToken, &whileStmt->openParenTkIdx_)
                 && parseExpression(whileStmt->cond_)
                 && match(CloseParenToken, &whileStmt->closeParenTkIdx_)
-                && parseStatement(whileStmt->stmt_));
+                && parseStatement(whileStmt->stmt_, stmtCtx));
 }
 
 /**
@@ -453,7 +455,8 @@ bool Parser::parseWhileStatement_AtFirst(StatementSyntax*& stmt)
  *
  * \remark 6.8.5.2
  */
-bool Parser::parseDoStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseDoStatement_AtFirst(StatementSyntax*& stmt,
+                                      StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == Keyword_do,
@@ -465,7 +468,7 @@ bool Parser::parseDoStatement_AtFirst(StatementSyntax*& stmt)
     doStmt->doKwTkIdx_ = consume();
 
     return checkStatementParse(
-                parseStatement(doStmt->stmt_)
+                parseStatement(doStmt->stmt_, stmtCtx)
                 && match(Keyword_while, &doStmt->whileKwTkIdx_)
                 && match(OpenParenToken, &doStmt->openParenTkIdx_)
                 && parseExpression(doStmt->cond_)
@@ -479,7 +482,8 @@ bool Parser::parseDoStatement_AtFirst(StatementSyntax*& stmt)
  *
  * \remark 6.8.5.3
  */
-bool Parser::parseForStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseForStatement_AtFirst(StatementSyntax*& stmt,
+                                       StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == Keyword_for,
@@ -578,7 +582,7 @@ bool Parser::parseForStatement_AtFirst(StatementSyntax*& stmt)
     }
 
     return checkStatementParse(
-                parseStatement(forStmt->stmt_));
+                parseStatement(forStmt->stmt_, stmtCtx));
 }
 
 /**
@@ -615,7 +619,8 @@ bool Parser::parseGotoStatement_AtFirst(StatementSyntax*& stmt)
  *
  * \remark 6.8.6.2
  */
-bool Parser::parseContinueStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseContinueStatement_AtFirst(StatementSyntax*& stmt,
+                                            StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == Keyword_continue,
@@ -636,7 +641,8 @@ bool Parser::parseContinueStatement_AtFirst(StatementSyntax*& stmt)
  *
  * \remark 6.8.6.3
  */
-bool Parser::parseBreakStatement_AtFirst(StatementSyntax*& stmt)
+bool Parser::parseBreakStatement_AtFirst(StatementSyntax*& stmt,
+                                         StatementContext stmtCtx)
 {
     DEBUG_THIS_RULE();
     PSYCHE_ASSERT(peek().kind() == Keyword_break,

--- a/C/syntax/SyntaxKind.h
+++ b/C/syntax/SyntaxKind.h
@@ -463,10 +463,11 @@ enum SyntaxKind : std::uint16_t
     AmbiguousCallOrVariableDeclaration,
     AmbiguousMultiplicationOrPointerDeclaration,
 
+    ENDof_Node = AmbiguousMultiplicationOrPointerDeclaration,
+
 //=================================================================== Misc
 
-    Unknown,
-    ENDof_Node = Unknown
+    UnknownSyntax
 };
 
 enum class SyntaxKindCategory : char

--- a/C/syntax/SyntaxNode.cpp
+++ b/C/syntax/SyntaxNode.cpp
@@ -462,7 +462,7 @@ std::string PSY_C_API to_string(SyntaxKind kind)
 
 //=================================================================== Misc
 
-        case Unknown:
+        case UnknownSyntax:
             return "<UNKNOWN>";
         default:
             return tokenNames[kind];

--- a/C/tests/TestParser_2000_2999.cpp
+++ b/C/tests/TestParser_2000_2999.cpp
@@ -997,7 +997,7 @@ void TestParser::case2352()
 {
     parseStatement(R"(
                    switch ( x ) {
-                       case 1: y ;
+                       case 1 : y ;
                        case 2 : z ( ) ;
                    }
                    )");
@@ -1012,7 +1012,7 @@ void TestParser::case2354()
 {
     parseStatement(R"(
                    switch ( x ) {
-                       case 1: y ;
+                       case 1 : y ;
                        default : z ( ) ;
                    }
                    )");
@@ -1041,7 +1041,16 @@ void TestParser::case2357()
                                                Parser::DiagnosticsReporter::ID_of_UnexpectedDefaultLabelOutsideSwitch));
 }
 
-void TestParser::case2358() {}
+void TestParser::case2358()
+{
+    parseStatement(R"(
+                   switch ( x ) {
+                       case 1 :
+                       case 2 : break;
+                   }
+                   )");
+}
+
 void TestParser::case2359() {}
 void TestParser::case2360() {}
 void TestParser::case2361() {}

